### DISCUSSION TITLE
Thru::Mode or MIDIUART.turnThruOff(); ??

### DISCUSSION THREAD

**Discussion by AaronHaCohen** [FortySevenEffects/arduino_midi_library#351] [Q&A]
Hello, I'm using the Callbacks.ino example as a starting point for processing MIDI in and out. The example works fine but I cannot turn off MIDI Thru. It seems that MIDIUART.turnThruOff(); is no longer in the MIDI library. Thru::Mode is listed as a structure current docs but wont compile tho I do <#include midi_Defs.h> and tried both Thru::Off, Thru::0 and Thru.Off and 0.
Can anyone help me to turn off MIDI Thru? I'm also curious why the scope resolution operator :: is shown to address a structure instead of the dot operator.  I confess I'm a newbee at C++ but have been writing ASM and plain C for many years and wrote a lot of MIDI processing code 40 years in DOS and a synth editor in Win98. Now I getting back into MIDI processing in my retirement and am slow catching up with C++  Thanks for any help. Aaron

**Comment by lathoub** [comment]
Hi `turnThruOff` should do the trick.
Can you post a minimized version of your code, so we can check

**Comment by AaronHaCohen** [comment]
Thanks for your suggestion. Yes, I found MIDI.turnThruOff();  In the library and tried it but it doesn't turn off MIDI thru. Any suggestions? I just posted details and the code. Thanks for your reply, Aaron

**Comment by lathoub** [comment]
It's not a good habit to attach code, it's better to post the code as part of the body and tag it with ``` (the `<>` button).
Also, what hardware and IDE are you using?

**Reply by AaronHaCohen** [reply]
Thanks again for the quick response and the suggestion. And sorry, I'm a newbee on this forum and learning how to use it. I'll see if I can succeed with the <> button. I'm using a Uno R3 on Arduino 2.3.2 on a Win 10 PC. I've written a fair amount of code with this setup without problems until now. I'd appreciate any help available. The code is the Callback.ino example with MIDI.turnThruOff(); added at line 40. It compiles, uploads and runs ok but the turnThruOff line seems to be ignored. Here goes an attempt to use the <> button:  Did it work?  Thanks again for your help, Aaron


`// Callbacks2.ino
// added MIDI.turnThruOff(); at line 43
#include <MIDI.h>

MIDI_CREATE_DEFAULT_INSTANCE();

// -----------------------------------------------------------------------------

// This function will be automatically called when a NoteOn is received.
// It must be a void-returning function with the correct parameters,
// see documentation here:
// https://github.com/FortySevenEffects/arduino_midi_library/wiki/Using-Callbacks

void handleNoteOn(byte channel, byte pitch, byte velocity) {
  // Do whatever you want when a note is pressed.

  // Try to keep your callbacks short (no delays ect)
  // otherwise it would slow down the loop() and have a bad impact
  // on real-time performance.
}

void handleNoteOff(byte channel, byte pitch, byte velocity) {
  // Do something when the note is released.
  // Note that NoteOn messages with 0 velocity are interpreted as NoteOffs.
}

// -----------------------------------------------------------------------------

void setup() {
  // Connect the handleNoteOn function to the library,
  // so it is called upon reception of a NoteOn.
  MIDI.setHandleNoteOn(handleNoteOn);  // Put only the name of the function

  // Do the same for NoteOffs
  MIDI.setHandleNoteOff(handleNoteOff);

  // Initiate MIDI communications, listen to all channels
  MIDI.begin(MIDI_CHANNEL_OMNI);
  // line 40 added 
  MIDI.turnThruOff();
}

void loop() {
  // Call MIDI.read the fastest you can for real-time performance.
  MIDI.read();

  // There is no need to check if there are messages incoming
  // if they are bound to a Callback function.
  // The attached method will be called automatically
  // when the corresponding message has been received.
}
`

**Reply by lathoub** [reply]
(just surround the code with ``` at the start and again ``` at the end)
Example
```
 #include <MIDI.h>
```

**Comment by lathoub** [comment]
Code looks good at first glance - assuming you are using a MIDI shield?

**Reply by AaronHaCohen** [reply]
Thanks. No I'm not using a MIDI shield. I hardwire directly to the UNO for reliability. Hardware is not a problem. I'm a retired EE and have designed and built a lot of signal processing equipment. Again, I appreciate your help. Aaron

**Comment by lathoub** [comment]
>  I'm a retired EE and have designed and built a lot of signal processing equipment.

Cool

Calling `MIDI.turnThruOff()` after `MIDI.begin(MIDI_CHANNEL_OMNI);` is the correct execution order

in `begin()` thru is set to Full and activated

https://github.com/FortySevenEffects/arduino_midi_library/blob/2d64cc3c2ff85bbee654a7054e36c59694d8d8e4/src/MIDI.hpp#L96-L97

`MIDI.turnThruOff()` then turns it off

https://github.com/FortySevenEffects/arduino_midi_library/blob/2d64cc3c2ff85bbee654a7054e36c59694d8d8e4/src/MIDI.hpp#L1440-L1441

So that part of the code is looking good, AFAIK this should work.

How do you know that Thru is still on?

**Reply by AaronHaCohen** [reply]
I really appreciate you digging into the problem like this. Tx of the UNO is connected to MIDI-OX a MIDI monitor. It shows every byte that the MIDI controller sends, which is connected to Rx. It sees Active sensing, NoteOn, NoteOff, CC+breath, and PitchBend; exactly the same as I get when connecting the controller directly to the monitor. The controller is a Yamaha Wx-11 sending standard MIDI at 5 volts. So it's TTL all the way.

The test code I posted is just a copy of the Callbacks.ino in the examples directory with only the addition of the MIDI.thruTurnOff.   Output to the MIDI monitor is exactly the same with or without that added line. I also previously put in some watchdog writes in the void handleNoteOn(...etc,  after line 14 to check if that handler is working. It is! The watchdog messages are sent along in addition to all the MIDI thru messages.

Yes, AFAIK it should be working fine! One limitation I have is that all my code writing over the last 60 years has been in ASM or plain C, and a little bit of Fortran. This C++ that all the kids are writing is quite a puzzlement to me at times. But I guess it might help me avoid old timer's disease and keep me off the streets.

But I'm baffled!  Thanks again, Aaron
